test/Microsoft.Framework.OptionsModel.Test/ConfigurationCollectionBindingTests.cs

```diff
@@ -0,0 +1,299 @@
+        {
+            var input = new Dictionary<string, string>
+            {
+                {"ObjectDictionary:abc:Integer", "1"},
+                {"ObjectDictionary:def:Integer", "2"},
+                {"ObjectDictionary:ghi:Integer", "3"}
+            };
+
+            var config = new Configuration(new MemoryConfigurationSource(input));
+
+            var binder = new ConfigurationBinder(config);
+            var options = binder.Read<OptionsWithDictionary>();
+
+            Assert.Equal(3, options.ObjectDictionary.Count);
+
+            Assert.Equal(1, options.ObjectDictionary["abc"].Integer);
+            Assert.Equal(2, options.ObjectDictionary["def"].Integer);
+            Assert.Equal(3, options.ObjectDictionary["ghi"].Integer);
+        }
+
+        [Fact]
+        public void ListDictionary()
+        {
+            var input = new Dictionary<string, string>
+            {
+                {"ListDictionary:abc:0", "abc_0"},
+                {"ListDictionary:abc:1", "abc_1"},
+                {"ListDictionary:def:0", "def_0"},
+                {"ListDictionary:def:1", "def_1"},
+                {"ListDictionary:def:2", "def_2"}
+            };
+
+            var config = new Configuration(new MemoryConfigurationSource(input));
+
+            var binder = new ConfigurationBinder(config);
+            var options = binder.Read<OptionsWithDictionary>();
+
+            Assert.Equal(2, options.ListDictionary.Count);
+            Assert.Equal(2, options.ListDictionary["abc"].Count);
+            Assert.Equal(3, options.ListDictionary["def"].Count);
+
+            Assert.Equal("abc_0", options.ListDictionary["abc"][0]);
+            Assert.Equal("abc_1", options.ListDictionary["abc"][1]);
+            Assert.Equal("def_0", options.ListDictionary["def"][0]);
+            Assert.Equal("def_1", options.ListDictionary["def"][1]);
+            Assert.Equal("def_2", options.ListDictionary["def"][2]);
+        }
+
+        [Fact]
+        public void NonStringKeyDictionaryBinding()
+        {
+            var input = new Dictionary<string, string>
+            {
+                {"NonStringKeyDictionary:abc", "val_1"},
+                {"NonStringKeyDictionary:def", "val_2"},
+                {"NonStringKeyDictionary:ghi", "val_3"}
+            };
+
+            var config = new Configuration(new MemoryConfigurationSource(input));
+
+            var binder = new ConfigurationBinder(config);
+            var options = binder.Read<OptionsWithDictionary>();
+
+            Assert.Null(options.NonStringKeyDictionary);
+        }
+
+        private class CustomList: List<string>
+        {
+            // Add an overload, just to make sure binding picks the right Add method
+            public void Add(string a, string b)
+            {
+            }
+        }
+
+        private class CustomDictionary<T>: Dictionary<string, T>
+        {
+        }
+
+        private class NestedOptions
+        {
+            public int Integer { get; set; }
```

Comment:
Can we also nest a List/Dictionary in here as well just to make sure that also works?